### PR TITLE
GH-41389: [Python] Expose byte_width and bit_width of ExtensionType in terms of the storage type

### PR DESCRIPTION
**Pull request by llama90**
### Rationale for this change

This update aligns the Python API with Arrow C++ by exposing the actual byte and bit widths of extension types from their storage type.

### What changes are included in this PR?

- Expose byte_width and bit_width properties for ExtensionType in Python, reflecting the underlying storage type.
- Add  unit tests to verify these properties

### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes

* GitHub Issue: #41389

### CODE REVIEW

File: python/pyarrow/tests/test_extension_type.py
Comment (kou) [inline python/pyarrow/tests/test_extension_type.py:270]:
```suggestion
def test_ext_type_byte_width():
```

File: python/pyarrow/tests/test_extension_type.py
Comment (kou) [inline python/pyarrow/tests/test_extension_type.py:277]:
```suggestion
def test_ext_type_bit_width():
```

**Review by kou — APPROVED** [review]
+1

File: python/pyarrow/tests/test_extension_type.py
Comment (llama90) [reply python/pyarrow/tests/test_extension_type.py]: When I looked at the function nearby, I understood that the notation (`__`) was a function for testing `@property`.

After reading the review and searching the whole thing, I realized it was a typo and modified the other functions I used as a reference.

Thank you for review!

**Comment by kou** [comment]
I'll merge this in a few days if nobody objects this.

**Review by jorisvandenbossche — COMMENTED** [review]
Can you add a test for the case where the storage type is _not_ a fixed width type (and so doesn't have those values defined)

File: python/pyarrow/tests/test_extension_type.py
Comment (llama90) [inline python/pyarrow/tests/test_extension_type.py:279]: @jorisvandenbossche 

Hello! I've added the `LabelType` as a non-fixed width type in tests. 

Is this aligned with the intended behavior for handling these types? Thanks!

**Comment by kou** [comment]
Could you rebase on main to fix CI failures?

I'll merge this in the next week if nobody objects this.

**Comment by llama90** [comment]
Hello @kou! Will this PR need more time before it can be merged?

Thank you as always!

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:275]:
```suggestion
    assert ty.byte_width == 16
    ty = ParamExtType(5)
    assert ty.byte_width == 5
```

All those tests should check the extension type itself, and not trough `.storage_type`, because the goal of the PR is to make those attributes available on the extension type itself?

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:293]:
```suggestion
        _ = ty.byte_width


def test_ext_type_bit_width():
    # Test for fixed-size binary types
    ty = UuidType()
    assert ty.bit_width == 128
    ty = ParamExtType(5)
    assert ty.bit_width == 40

    # Test for non fixed-size binary types
    ty = LabelType()
    with pytest.raises(ValueError, match="Non-fixed width type"):
        _ = ty.byte_width
```

**Review by jorisvandenbossche — COMMENTED** [review]
Apologies for the slow follow-up. Spotted one more issue with the tests

File: python/pyarrow/tests/test_extension_type.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_extension_type.py:279]: I think we want that the extension type is consistent here with the normal data types, which raise an error in this case (you can check how the error is raised in the base DataType class)

**Comment by llama90** [comment]
@jorisvandenbossche Fine! I applied your suggestions and fixed the unit test.

Thank you for review :)

File: python/pyarrow/tests/test_extension_type.py
Comment (llama90) [reply python/pyarrow/tests/test_extension_type.py:279]: You mean that the extension type should make erorr? In this case, the extension type return -1.

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

**Pull request merged**
